File: src/Library/ObjectModel/Genre.vb
Public Class Genre
    Inherits Pixa.Soundbridge.Library.MediaObject

    Private _songs As SongCollection

    Friend Sub New(ByVal server As MediaServer, ByVal name As String)
        MyBase.New(server, name)
    End Sub

    Public ReadOnly Property Songs() As SongCollection
        Get
            If _songs Is Nothing OrElse Not _songs.IsActive Then
                _songs = New SongCollection(Server.Soundbridge)
                Dim r As String = Client.SetBrowseFilterGenre(Name)
                If r <> "OK" Then ExceptionHelper.ThrowCommandReturnError("SetBrowseFilterGenre", r)

                Dim songNames As String() = Client.ListSongs
                If songNames.Length = 1 AndAlso (songNames(0) = "ErrorDisconnected" Or songNames(0) = "GenericError") Then ExceptionHelper.ThrowCommandReturnError("ListSongs", songNames(0))

                For i As Integer = 0 To songNames.Length - 1
                    _songs.Add(New Song(Server, songNames(i), i))
                Next
            End If

            Return _songs
        End Get
    End Property
End Class

Public Class GenreCollection
    Inherits Pixa.Soundbridge.Library.SoundbridgeObjectCollection(Of Genre)

    Friend Sub New(ByVal sb As Soundbridge)
        MyBase.New(sb)
    End Sub
End Class

File: src/Pixa.Soundbridge/Genre.cs
namespace Pixa.Soundbridge
{
    public class Genre : MediaObject
    {
        private SongCollection _songs;

        internal Genre(MediaServer server, int index, string name) : base(server, index, name)
        {
        }

        public SongCollection Songs
        {
            get
            {
                if (_songs is null || !_songs.IsActive)
                {
                    _songs = new SongCollection(Server.Soundbridge);
                    string r = Client.SetBrowseFilterGenre(Name);
                    if (r != "OK")
                        ExceptionHelper.ThrowCommandReturnError("SetBrowseFilterGenre", r);
                    var songNames = Client.ListSongs();
                    if (songNames.Length == 1 && songNames[0] == "ErrorDisconnected" | songNames[0] == "GenericError")
                        ExceptionHelper.ThrowCommandReturnError("ListSongs", songNames[0]);
                    for (int i = 0, loopTo = songNames.Length - 1; i <= loopTo; i++)
                        _songs.Add(new Song(Server, int.Parse(songNames[i]), i.ToString()));
                }

                return _songs;
            }
        }
    }
}